rust/crates/sdk/src/stream.rs

```diff
@@ -0,0 +1,286 @@
+        });
+
+        let conn = connect(config, &feed_ids, stats.clone()).await?;
+
+        let water_mark = Arc::new(Mutex::new(HashMap::new()));
+
+        Ok(Stream {
+            config: config.clone(),
+            feed_ids,
+            conn: Some(conn),
+            report_sender,
+            report_receiver,
+            shutdown_sender,
+            stats,
+            water_mark,
+        })
+    }
+
+    /// Starts listening for reports on the Stream.
+    /// This method will spawn a new task for each WebSocket connection.
+    pub async fn listen(&mut self) -> Result<(), StreamError> {
+        let conn = self
+            .conn
+            .take()
+            .ok_or_else(|| StreamError::ConnectionError("No connection".into()))?;
+
+        match conn {
+            WebSocketConnection::Single(stream) => {
+                let report_sender = self.report_sender.clone();
+                let shutdown_receiver = self.shutdown_sender.subscribe();
+                let stats = self.stats.clone();
+                let water_mark = self.water_mark.clone();
+                let config = self.config.clone();
+                let feed_ids = self.feed_ids.clone();
+
+                tokio::spawn(run_stream(
+                    stream,
+                    report_sender,
+                    shutdown_receiver,
+                    stats,
+                    water_mark,
+                    config,
+                    feed_ids,
+                ));
+            }
+            WebSocketConnection::Multiple(streams) => {
+                for stream in streams {
+                    let report_sender = self.report_sender.clone();
+                    let shutdown_receiver = self.shutdown_sender.subscribe();
+                    let stats = self.stats.clone();
+                    let water_mark = self.water_mark.clone();
+                    let config = self.config.clone();
+                    let feed_ids = self.feed_ids.clone();
+
+                    tokio::spawn(run_stream(
+                        stream,
+                        report_sender,
+                        shutdown_receiver,
+                        stats,
+                        water_mark,
+                        config,
+                        feed_ids,
+                    ));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Reads the next available report on the Stream.
+    /// Reads blocks until a report is received, the context is canceled or all underlying connections are in a error state.
+    ///
+    /// # Returns
+    ///
+    /// * `WebSocketReport` - The next available report.
+    pub async fn read(&mut self) -> Result<WebSocketReport, StreamError> {
+        self.report_receiver
+            .recv()
+            .await
+            .ok_or(StreamError::StreamClosed)
+    }
+
+    /// Closes the Stream.
+    /// It is the caller's responsibility to call close when the stream is no longer needed.
```

Comment:
I don't know if it's desirable, but worst case you can block in the `Drop` impl to perform any cleanup that may have been missed.